### ISSUE TITLE
[fluxv2] Allow some helm repositories in cluster to be unmanaged (un-exposed/private/intrenal) by kubeapps

### ISSUE THREAD

**Issue by hongkunyoo** [kubeapps/kubeapps#4298]
### Description:

~~It would be great to have flux type AppRepository feature.~~
It would be great to have a mechanism to make some `HelmRepository` as unmanaged.

### Background

I am currently planning to use fluxv2 plugin only.
There are customers using our cluster as clients.
I am the cluster admin who provides kubeapps.

### Current behavior

It seems fluxv2 searches`HelmRepository` from all namespaces.
So it needs Cluster Scope permission of listing `HelmRepository` cluster-wide

### Feature request

It would be great if kubeapps admin can choose which `HelmRepository` to provide to users.

### Reason

There are several `HelmRepositories` in our cluster.
Some are for internal use only, some are for our customers using our cluster.
So I want to separate the `HelmRepository` based on use cases.

Is it possible to explicitly choose which `HelmRepository` to expose to kubeapps users instead of exposing all `HelmRepository` registered in our cluster?

Once again, thank you for providing such a wonderful project!

**Comment by absoludity** [comment]
Thanks for testing out the flux support @hongkunyoo ! @gfichtenholt , let me know if you're keen to work through the options here with @hongkunyoo .

**Comment by gfichtenholt** [comment]
yes, no problem. I am on PTO all next week though, so it'll have to be after that.

**Comment by absoludity** [comment]
@hongkunyoo Just for some background, Greg will be starting/continuing work on the app repository API when back. You can read some of the background and the current plan for the [protobuf messages in the relevant section of the design doc](https://docs.google.com/document/d/1z03msZRGsRvcRaom-yrvEwIWcEfNy6fSA5Zg28gjYvA/edit#heading=h.u3lj7whdof7t). (It should be setup so anyone can comment). The first PR for this work was proposed at the end of last week - see #3496

**Comment by hongkunyoo** [comment]
Thank you for sharing the status. I am really looking forward to having this feature!
Always thankful for your work!

**Comment by gfichtenholt** [comment]
Hello,
A few comments:

re: 
> It seems fluxv2 searches HelmRepository from all namespaces

Yes. To be specific, it is fluxv2 kubeapps plugin that **chooses** to look for helm repositories in all namespaces. In other words, I wrote code in out plugin to look for repositories in all namespaces, mostly because flux itself allows one to essentially use a repository from any namespace, provided the service account on behalf of which kubeappsapis deployment is running (kubeapps-internal-kubeappsapis) has appropriate RBAC settings to access namespaces. Remember, unlike existing helm support, I don't really have a database where I persist repositories added by a user. My ultimate source of truth is k8s cluster, so I ask for all of it (today)

> So it needs Cluster Scope permission of listing HelmRepository cluster-wide

To be very specific, at this time there is a cluster role `kubeapps:controller:kubeapps-apis-fluxv2-plugin` and a 
cluster role binding `kubeapps:controller:kubeapps-apis-fluxv2-plugin` which is what determines which helm repositories can be listed. In think at least in dev environment, its `cluster-admin` access, which means *everything*. Not sure about what production setup looks like. @absoludity probably knows more about that.

> It would be great if kubeapps admin can choose which HelmRepository to provide to users. 

I have several thoughts on possible approaches:

1) Reading the documentation (https://kubernetes.io/docs/reference/access-authn-authz/rbac/) it sure looks like k8s RBAC is flexible enough to actually let you specify which resource names that one is allowed access to. Disclaimer: I have not yet tried this approach, and it does look somewhat cumbersome to maintain up-to-date. Like every resource name would need to be listed. If you'd like I can spend some time testing out the limits of this with our flux support. The good thing about this approach is it would not need any code changes for us.

If (1) is not sufficient or does not look like what you'd want to use then I believe we could work out some approach in kubeapps. I think any solution would require new configuration, as there is no place to hang it today. I am thinking flux plug-in specific configuration configmap or a file (which we already have but just a couple of minor things). To decide exactly what is missing, I'd like to know:

2) Are you interested in a "whitelist" (opt-in) approach or "blacklist" (opt-out) approach, i.e. 
do you want to be able to say
 - "repository x and y" are public and everything else is NOT public. That is kind of similar to the way 
    things work with helm support today but not exactly, because unlike helm, with flux I have to take into account repositories
    added by kubeapps via UX/our API or simply added to the cluster via flux CLI or kubectl CLI, bypassing us completely. So I am not always in control of the process when repositories are added OR
 - "repository x and y" are NOT public and everything else is public

3) Would you want a solution that actually lets you be very fine grained, like choose a specific set of helm repositories to be accessible to kubeapps users, or a little more coarse grained, like maybe only repositories in certain namespaces? Each solution has pluses and minuses. 

Thanks 
Greg

**Comment by gfichtenholt** [comment]
@ppbaena, what is the priority of this in relation to the rest of #3496

**Comment by hongkunyoo** [comment]
Thank you for replying my request. I really appreciate it.

For more details about my use cases. Let's me explain my situation. And please let me know if my explanation is insufficient(not a native speaker).

I am using OIDC integration with Kubernetes as well as Kubeapps login using authProxy feature.
First I thought the `HelmRepository` will be listed by following K8s RBAC settings. So I tried but found out it didn't work since fluxv2 plugin searches from all namespaces.
However, soon enough I realized that it was hard to find out which namespace to list `HelmRepository` because as you mentioned:

> unlike existing helm support, I don't really have a database where I persist repositories added by a user.

So I thought maybe using AppRepository to notify fluxv2 plugin where to look in.

-----

If the (1) approach works, it will be the best. if it does not work, I would want a fine grained approach to choose which `HelmRepository` to expose.

------

In my case, giving a clusterrole of viewing `HelmRepository` permission to view from all namespaces to users is not an issue.
I just want users to view specific `HelmRepository` from **kubeapps dashboard**.

So, If using RBAC to have a fine grained control of viewing specific `HelmRepository` in kubeapps dashboard is not feasible, using different mechanism is also fine with me.

For example, using annotation to decide which `HelmRepository` to be expose.
Then fluxv2 plugin can search all namespaces as usual, and just filter out some unwanted `HelmRepository`.

```yaml
apiVersion: source.toolkit.fluxcd.io/v1beta1
kind: HelmRepository
metadata:
  annotations:
     kubeapps-dashboard-expose: true   # <-- just an example
  name: bitnami
  namespace: flux-system
spec:
  interval: 10m0s
  timeout: 1m0s
  url: https://charts.bitnami.com/bitnami
---
apiVersion: source.toolkit.fluxcd.io/v1beta1
kind: HelmRepository
metadata:
  annotations:
     kubeapps-dashboard-expose: false   # <-- do not expose
  name: internal-repo
  namespace: flux-system              # <-- does not matter whether it is in same namespace or not.
spec:
  interval: 10m0s
  timeout: 1m0s
  url: https://charts.internal-repo.com
```

**Comment by gfichtenholt** [comment]
Annotations could work, let me think about it some. It's very coarse grained, but maybe sufficient for MVP. The use case it doesn't handle is when you want to be able to give user 1 access to repo A and repo B, while user2 can only access repo A

FWIW, There are a couple of other approaches that came to mind and I wanted to mention them for posterity:

1a) we can separate the list of namespaces that flux plugin searches from the ones that the currently logged in user sees. E.g. suppose there are two repositories A in ns1 and B in ns2. flux plugin background process would read/index both repos, but suppose the currently logged in user only has access to ns1 via RBAC settings. Then they would only see repo A in the list. This would be a relatively simple change to implement. Again, that requires the admin to actively manage RBAC settings for all the users in the cluster
1b) We can make a change so that while flux plugin reads/indexes all the repos, the logged user only sees  the ones from the current namespace selected in the dashboard context. This could be enabled/disable via some flag in flux plugin configurarion

Also, I should mention that whatever change we make in kubeapps, including your proposal with annotations, isn't going to stop the user from accessing flux repos via flux CLI or kubectl CLI thus bypassing kubeapps

**Comment by hongkunyoo** [comment]
> Are you open to a solution that only filters out repos based on the namespace they are in? In your example both repos are in the same namespace so this wouldn't work.

Namespace does not matter. If only I could control the visibility of repository viewing from kubeapps dashboard, anything will be fine with me: AppRepository, namespace and annotations are just one of many suggestions.

> 1b) We can make a change so that while flux plugin reads/indexes all the repos, the logged user only sees the ones from the current namespace selected in the dashboard context. This could be enabled/disable via some flag in flux plugin configurarion

In my case, this approach would be less attractive, since I have to add every `HelmRepository` to every namespace that users use if certain `HelmRepository` is in common use.(Background explanation: We have lots of tenant namespaces and number of namespace can keep grow) However for fine grained control, this might be the best, since it covers the whole use cases.

If I could be able to choose which namespace to expose, I was going to make `public-repo` namespace(for example) and let all kubeapps users to view that namespace: This is why I preferred whitelist approach over blacklist.

> isn't going to stop the user from accessing flux repos via flux CLI or kubectl CLI thus bypassing kubeapps

Yes, totally I agree with you.

**Comment by gfichtenholt** [comment]
I need to clarify something. You keep using the word 'dashboard'. FYI, dashboard uses gRPC API to talk to kubeapps backend. It's just one way to interact with kubeapps. One can invoke kubeapps via grpc client directly or even via HTTP.  One day we may add a kubeapps CLI as another way to interfact with kubeapps backend. All this to say that rather than use a term like `kubeapps-dashboard-expose` I would use something like `managed-by-kubeapps`. Also, if a repo is not "managed" by kubeapps, kubeapps need not even be aware of it. I see no point in using server resources like memory or CPU to waste on processing repositories not managed by kubeapps. 

Also, your last comment is a little bit confusing. You say 'Namespace does not matter' as if to say the approach based on filtering by namespace is not acceptable. Then, later you say 'If I could be able to choose which namespace to expose, I was going to make public-repo namespace(for example) and let all kubeapps users to view that namespace'. That sounds like you're saying the namespace-based approach is acceptable. Could you clarify?
Thanks

**Comment by gfichtenholt** [comment]
One more question to help me make a decision: if a repo is not "managed" by kubeapps, what is the point of even adding it to a cluster? Are you planning to use flux CLI or kubectl CLI to interact with it?

**Comment by hongkunyoo** [comment]
> One more question to help me make a decision: if a repo is not "managed" by kubeapps, what is the point of even adding it to a cluster? Are you planning to use flux CLI or kubectl CLI to interact with it?

Yes, exactly.

We want users to utilize Kubernetes either by `kubectl` or `kubeapps`.
Users who are familiar with YAML manifests might use `kubectl` directly, and users who are beginners to K8s might use `kubeapps` dashboard.

We are thinking of  `kubectl`, `helm`, `kubeapps` are one of many ways to access Kubernetes. Similar to your plan.
> One can invoke kubeapps via grpc client directly or even via HTTP. One day we may add a kubeapps CLI as another way to interfact with kubeapps backend

**Comment by gfichtenholt** [comment]
So is there ever a scenario of a user adding a flux helm repo via flux CLI or kubectl CLI such that that same user or another user uses kubeapps to interact with it (e.g. list/install packages)?

**Comment by hongkunyoo** [comment]
Currently, we are planning **not** to give flux helm repo **create** permission to users.
Only admin can add flux helm repo.

If users want more repo, they need to request to us.

The reason why I said "currently" is because if there are many requests that users want to add their own flux helm repo, we might think about giving create permission. However currently, we do not think there will be such many cases.

**Comment by gfichtenholt** [comment]
So an admin (i.e. also a user with elevated privileges) might add a flux helm repo via kubectl. And then another user might wants to use kubeapps to list/install packages from this repo?

**Comment by gfichtenholt** [comment]
OK, thanks, I think I got it now. It sounds like the namespace-based approach is not the right fit. Bottom line is there will be repos not "managed" by kubeapps at all, so kubeapps need not even know about them. There will also be be repos "managed" by kubeapps and thus visible though dashboard, gRPC and available for interaction. Whether or not they can be in the same namespace is up to the design.
It sounds like we basically need a way to flag a given repo as "managed" or not. Whether that is by placing it in a certain  whitelisted namespace or an annotation on the object is up to the design. Did I get this right?
Thanks

**Comment by hongkunyoo** [comment]
Yes, It is perfect!

> It sounds like we basically need a way to flag a given repo as "managed"

I think this is what I wanted.

Thank you for listening to my use case!

**Comment by gfichtenholt** [comment]
Thank you! We will have some discussions internally and get back to you

**Comment by absoludity** [comment]
Hi @hongkunyoo . Greg and I just talked about the different options here, and I wanted to clarify some current behavior related to your request.

1. Currently, users who log in to Kubeapps will see packages from all Flux `HelmRepositories` on the cluster, even if the user does not have "read" permission to see the `HelmRepository` via the CLI. This is an issue which @gfichtenholt is already aware of and planning to fix. This issue results in the situation you describe: users see charts from `HelmRepositories` across all namespaces on the cluster, even when they should not be reading them.
2. Once (1) is fixed, users who log in to Kubeapps will only see packages from the Flux `HelmRepositories` to which they have access (ie. if they have "read" permission to see the `HelmRepository` via the CLI). So users will not see `HelmRepositories` in Kubeapps if they cannot see them via the CLI. So an Admin user can add HelmRepositories in specific namespaces or with specific permissions so that users won't see them.

So once (2) is implemented, it would be possible for an admin to ensure users cannot see certain `HelmRepositories`, based on the existing cluster permissions (eg. RBAC). This is what Kubeapps should (and will) do once the issue is fixed, since it aims to be a user-interface showing the same results as the CLI.

Does that solve your feature request, or do you additionally need:

3. the ability to annotate some `HelmRepositories` so that Kubeapps won't display them to users, even if the logged-in user has access to them via the CLI?

We will definitely be doing (2), but are uncertain if (3) is additionally required, as it goes against Kubeapps' aim to show exactly what the CLI would show. If (3) is required, why would the admin not ensure the user cannot see these `HelmRepositories` via the CLI? (ie. ensuring the users do not have permission to "read" the `HelmRepositories`). We can do (3) if it is required, but want to be sure we have understood correctly.

Thanks for the input!

**Comment by hongkunyoo** [comment]
Thank you for considering my case.

I think option (2) will work for me, but with little minor issues.

Previously @gfichtenholt tell me about using current namespace selected in the dashboard context.

> 1b) We can make a change so that while flux plugin reads/indexes all the repos, the logged user only sees the ones from the current namespace selected in the dashboard context. This could be enabled/disable via some flag in flux plugin configurarion 

In this case, I have to create the same `HelmRepository` across all namespace when it's a common `HelmRepository`.(We use namespace as user tenant, so every different namespace has different user groups: namespace based multi-tenant cluster.)

if there is a way that admin can choose a default namespace that every user will automatically fall in when entering "Catalog" menu(for example, every user will automatically fall into `public-repo` namespace when entering "Catalog" menu) then, I can grant ever tenant user a "read" permission of `public-repo` namespace and create `HelmRepository` once in that namespace.

Or If user can see charts from `HelmRepository` without any action(Selecting which namespace to view `HelmRepository`), it will be the best.

I think my use case is somewhat in the middle of "Allowing to view from all namespace" and "Fine-grained view permission". --> maybe it seems to be an edge case?

-----

I understand complying with K8s RBAC mechanism seems more cohesive. 
Maybe I can deal with by just making a note for user to select `public-repo` manually in "Catalog" menu.

Long story short, I might need to test myself with option (2) whether it's enough for me and my tenant users.

i really appreciate for your consideration.

**Comment by absoludity** [comment]
> Previously @gfichtenholt told me about using current namespace selected in the dashboard context.
> 
> > 1b) We can make a change so that while flux plugin reads/indexes all the repos, the logged user only sees the ones from the current namespace selected in the dashboard context. This could be enabled/disable via some flag in flux plugin configurarion

That was just Greg brain-storming one possibility, while trying to make sure we understand your request I think? (ie. it's not how it's currently intended to work).

The normal behavior would be that the logged-in user sees all packages that they can install in the current namespace (ie. usually packages from all repos they can read, since flux allows you to create a `HelmRelease` in namespace A from a repo that you have access to in namespace B).

> if there is a way that admin can choose a default namespace that every user will automatically fall in when entering "Catalog" menu(for example, every user will automatically fall into `public-repo` namespace when entering "Catalog" menu) then, I can grant every tenant user a "read" permission of `public-repo` namespace and create `HelmRepository` once in that namespace.

Correct, that is the desired behavior. If all users can read that `HelmRepository`, then they see the packages in the catalog.

> 
> Or If user can see charts from `HelmRepository` without any action(Selecting which namespace to view `HelmRepository`), it will be the best.

Yes, with the scenario you describe, they will see the same catalog of charts regardless of which namespace they choose (the choice of namespace is where the package will be installed, not where the catalog comes from necessarily).

> I understand complying with K8s RBAC mechanism seems more cohesive. Maybe I can deal with by just making a note for user to select `public-repo` manually in "Catalog" menu.

I don't think this will be necessary, as above, they'll see the catalog - the namespace selection is the target for the installation (where the chart will be installed, not where it comes from).

> i really appreciate for your consideration.

Thanks for all the clarifications! @gfichtenholt , let me know if I've missed something above or if something needs correcting.

**Comment by gfichtenholt** [comment]
> That was just Greg brain-storming one possibility, while trying to make sure we understand your request I think? (ie. it's not how it's currently intended to work).

100% Correct, I was just exploring possibilities

**Comment by absoludity** [comment]
@gfichtenholt Based on the above, and your current work, I think this one is already in progress right? (as in, the PRs you're working on to ensure people see a catalog based on the repos they have access to will fix this, based on our understanding above?)

**Comment by gfichtenholt** [comment]
> @gfichtenholt Based on the above, and your current work, I think this one is already in progress right? (as in, the PRs you're working on to ensure people see a catalog based on the repos they have access to will fix this, based on our understanding above?)

Well, I've done the work to implement option (2) described above in https://github.com/kubeapps/kubeapps/issues/4298#issuecomment-1061296805. I haven't done any work for option (3), if indeed that is what @hongkunyoo  wants to see. I don't see a definitive answer one way or another from them

**Comment by absoludity** [comment]
Duplicate of #4390 . Closing as Greg finished this work with the PR's linked to that issue. Thanks Greg!

**Comment by absoludity** [comment]
Duplicate of #4390
